File: algorithms/1_easy/C/00070_climbing_stairs/dynamic_programming.py
class Solution:
    def climbStairs(self, n: int) -> int:
        # Edge case: Directly return 1 to reduce complexity of creating the memoization vector
        if n == 1:
            return 1

        mem = [0] * n

        mem[n - 1] = 1
        mem[n - 2] = 2

        for i in range(n - 3, -1, -1):
            mem[i] = mem[i + 1] + mem[i + 2]

        return mem[0]

File: algorithms/1_easy/C/00070_climbing_stairs/dynamic_programming.rs
impl Solution {
    pub fn climb_stairs(n: i32) -> i32 {

        // Edge case: Directly return 1 to reduce complexity of creating the memoization vector
        if n == 1 {
            return 1;
        }

        let n_usize = n as usize;
        let mut mem = vec![0_i32; n_usize];

        mem[n_usize - 1] = 1;
        mem[n_usize - 2] = 2;

        for i in (0..(n_usize - 2)).rev() {
            mem[i] = mem[i + 1] + mem[i + 2];
        }

        mem[0]
    }
}
